List WI - filter by Iterations

pranavgore09 opened the pull request:
Fixes https://github.com/fabric8io/fabric8-planner/issues/704 and https://github.com/fabric8io/fabric8-planner/issues/703

* List workitems API updated to have filter by iteration
* Updated tests to consume above changes
* Added WorkItemsByIteration URL in related links of list iterations
* Added test case to verify list of work items filter by iteration

pranavgore09 commented:
@aslakknutsen @kwk @baijum Can you please check this PR ?

pranavgore09 commented:
[test]

pranavgore09 commented:
Facing errors in build - 
```
[0;31m[ERROR][mUpdate failed for github.com/jteeuwen/go-bindata: Unable to get repository
.
.
.
[0;31m[ERROR]	[mFailed to set version on github.com/jteeuwen/go-bindata to bbd0c6e271208dce66d8fda4bc536453cd27fc4a: open /tmp/go/src/github.com/almighty/almighty-core/vendor/github.com/jteeuwen/go-bindata: no such file or directory
.
.

```

Ritsyy commented:
@pranavgore09 you can use gin index, it will make the search better, it's been done for filtering by assignee https://github.com/aslakknutsen/almighty-core/blob/320a30e34efab4e8e7c795fa658e1de4a544006a/migration/sql-files/014-wi-fields-index.sql

pranavgore09 commented:
@Ritsyy Yeah, but isn't that is already applied on the column `fields` ? And iteration lies there only.

Ritsyy commented:
@pranavgore09 oh yes, sorry my bad, yes the is index for fields column.

pranavgore09 commented:
@Ritsyy No problem. Thanks for taking a look and pointing that out ! Much appreciated.

Ritsyy commented:
@pranavgore09 your welcome :)

Ritsyy commented:
[test]

kwk commented on design/workitems.go:
```diff
@@ -96,6 +96,7 @@ var _ = a.Resource("workitem", func() {
 			a.Param("page[offset]", d.String, "Paging start position")
 			a.Param("page[limit]", d.Integer, "Paging size")
 			a.Param("filter[assignee]", d.String, "Work Items assigned to the given user")
+			a.Param("filter[iteration]", d.String, "Work Items which are in the given iteration")
```
The description text describes the response but not the parameter.

pranavgore09 replied:
@kwk  if I understand your comment correctly, I will change it to - `IterationID to filter Work Items`.
Right ?

kwk commented:
@pranavgore09 Please stop rebasing or amending your commit and instead add new ones, okay?

pranavgore09 commented:
@kwk Okay, I will try not to ammend or rebase. But what I have observed is - If you add a comment on line number 30 and if I make change on line number 30 then it will be shown in `"Show outdated"` (rebase or ammend does not matter here).
Also it becomes very difficult/headache if some else `rebase` or `squash` thrier branch or upstream and if I pull that branch.
Sorry @kwk for inconvenience..

kwk commented on workitem_blackbox_test.go:
```diff
@@ -900,13 +913,47 @@ func (s *WorkItem2Suite) TestWI2ListByAssigneeFilter() {
+
+	_, list := test.ListWorkitemOK(s.T(), s.svc.Context, s.svc, s.wi2Ctrl, nil, nil, &iterationID, nil, nil)
+	assert.Len(s.T(), list.Data, 1)
+	assert.Equal(s.T(), iterationID, *list.Data[0].Relationships.Iteration.Data.ID)
```
You should make the line before require.Len because your program shall stop before it crashes due to accessing the first element of an empty array.

pranavgore09 commented:
@kwk thanks for the input. Yes https://github.com/fabric8io/fabric8-planner/issues/704 is not updated to say that I am switching the logic of `filter[iteration]`.

Added [a comment](https://github.com/fabric8io/fabric8-planner/issues/704#issuecomment-275102878) and updated description of the issue to fix it.
